biz.aQute.repository/src/aQute/bnd/repository/maven/provider/Crawler.java

```diff
@@ -0,0 +1,83 @@
+package aQute.bnd.repository.maven.provider;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.osgi.util.promise.Promise;
+import org.osgi.util.promise.PromiseFactory;
+
+import aQute.bnd.http.HttpClient;
+import aQute.libg.glob.Glob;
+
+class Crawler {
+	private final static Pattern	HREF_P	= Pattern.compile("href\\s*=\\s*\"(?<uri>[^\"]+)\"",
+		Pattern.CASE_INSENSITIVE);
+
+	final PromiseFactory			pf;
+	final HttpClient				client;
+
+	public Crawler(HttpClient client, PromiseFactory pf) {
+		this.client = client;
+		this.pf = pf;
+	}
+
+	public List<URI> getURIs(URI uri, Predicate<URI> predicate) throws Exception {
+		assert uri != null;
+		assert predicate != null;
+
+		String page = client.build()
+			.get(String.class)
+			.go(uri);
+
+		List<Promise<List<URI>>> promises = new ArrayList<>();
+		List<URI> uris = new ArrayList<>();
+
+		Matcher m = HREF_P.matcher(page);
+		while (m.find()) {
+			String path = m.group("uri");
+			URI next = uri.resolve(path)
+				.normalize();
+
+			if (next.getPath()
+				.endsWith("/")) {
+
+				URI relativize = uri.relativize(next);
+				if (relativize.isAbsolute())
+					continue; // not recursive, points outside
+
+				Promise<List<URI>> downstream = pf.submit(() -> getURIs(next, predicate));
+				promises.add(downstream);
+			} else {
+				if (predicate.test(next))
+					uris.add(next);
+			}
+		}
+		for (Promise<List<URI>> p : promises) {
+			uris.addAll(p.getValue());
+		}
+		return uris;
+	}
+
+	public List<URI> getURIs(URI uri) throws Exception {
+		return getURIs(uri, u -> true);
+	}
+
+	public static Predicate<URI> predicate(String include, String exclude) {
```

Comment:
This should use the new `PathSet.find()` methods to build the predicate from a set of includes and excludes. I just added these in https://github.com/bndtools/bnd/pull/3740/commits/5897aaf42ae697b01f5adea3f966d158ca726c38#diff-e731aefa8d049ad50ec427de5c0f25e7R175-R207 so you will need to rebase on master.

Changing to use PathSet, means you wont need to change to add Glob.NONE.

Reply:
Not completely clear how it works, you can fix that if you want.